metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesFinder.scala

```diff
@@ -0,0 +1,109 @@
+package scala.meta.internal.metals.testProvider
+
+import scala.meta.internal.implementation.ImplementationProvider
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.metals.DefinitionProvider
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.debug.BuildTargetClasses
+
+import ch.epfl.scala.bsp4j.BuildTarget
+
+trait TestSuitesFinder {
+  def findTestSuites(): Seq[TestDiscovery]
+}
+
+final class TestSuitesFinderImpl(
+    buildTargets: BuildTargets,
+    buildTargetClasses: BuildTargetClasses,
+    definitionProvider: DefinitionProvider,
+    implementationProvider: ImplementationProvider
+) extends TestSuitesFinder {
+
+  override def findTestSuites(): Seq[TestDiscovery] = {
+    buildTargets.allBuildTargetIds.toList
+      .flatMap(buildTargets.info)
+      .filterNot(_.getDisplayName.endsWith("build"))
+      .map(buildTarget => createTestDiscoveryFor(buildTarget))
+      // don't send empty results
+      .filter(_.discovered.asScala.exists(_.nonEmpty))
+  }
+
+  private def createTestDiscoveryFor(buildTarget: BuildTarget) = {
+    val classes = buildTargetClasses
+      .classesOf(buildTarget.getId)
+      .testClasses
+      // filter out symbols which don't represent classes
+      .filter { case (symbol, _) => symbol.endsWith("#") }
+      .toList
+      .map { case (symbol, className) =>
+        createTestEntry(buildTarget, symbol, className)
+      }
+
+    TestDiscovery(
+      buildTarget.getDisplayName,
+      buildTarget.getId.getUri,
+      TestProviderImpl.groupTestsByPackage(classes).asJava
+    )
+  }
+
+  private def createTestEntry(
+      buildTarget: BuildTarget,
+      symbol: String,
+      className: String
+  ): TestEntry = {
+    val location = for {
+      definition <- definitionProvider
+        .toPath(symbol, List(buildTarget.getId))
+      location <- definition.toResult.flatMap(
+        _.locations.asScala.toList
+          .filter(_.getUri == definition.uri)
+          .headOption
+      )
+    } yield location
+    val fullyQualifiedName = className.split('.').toList
+    val testClass = TestDiscovery.TestSuite(
+      className,
+      fullyQualifiedName.takeRight(1).head,
```

Comment:
Do we have a guarantee here that this will never be Nil?

Reply:
Yes, `fullyQualifiedName` contains package + class name, so the wort case scenario is when package is empty, but there's still class name and this is exactly what this piece of code is looking for. 
Some renaming will be very helpful, for example `val fullyQualifiedName = className` looks very suspicious for me now.

Reply:
Now it should be clearer